// chordsheetjs@6.1.0/lib/key.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _note = _interopRequireDefault(require("./note"));

var _constants = require("./constants");

var _enharmonicNormalize = _interopRequireDefault(require("./normalize_mappings/enharmonic-normalize"));

var _MODIFIER_TRANSPOSITI;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _classPrivateMethodInitSpec(obj, privateSet) { _checkPrivateRedeclaration(obj, privateSet); privateSet.add(obj); }

function _checkPrivateRedeclaration(obj, privateCollection) { if (privateCollection.has(obj)) { throw new TypeError("Cannot initialize the same private elements twice on an object"); } }

function _classPrivateMethodGet(receiver, privateSet, fn) { if (!privateSet.has(receiver)) { throw new TypeError("attempted to get private field on non-instance"); } return fn; }

function _wrapRegExp() { _wrapRegExp = function _wrapRegExp(re, groups) { return new BabelRegExp(re, void 0, groups); }; var _super = RegExp.prototype, _groups = new WeakMap(); function BabelRegExp(re, flags, groups) { var _this = new RegExp(re, flags); return _groups.set(_this, groups || _groups.get(re)), _setPrototypeOf(_this, BabelRegExp.prototype); } function buildGroups(result, re) { var g = _groups.get(re); return Object.keys(g).reduce(function (groups, name) { return groups[name] = result[g[name]], groups; }, Object.create(null)); } return _inherits(BabelRegExp, RegExp), BabelRegExp.prototype.exec = function (str) { var result = _super.exec.call(this, str); return result && (result.groups = buildGroups(result, this)), result; }, BabelRegExp.prototype[Symbol.replace] = function (str, substitution) { if ("string" == typeof substitution) { var groups = _groups.get(this); return _super[Symbol.replace].call(this, str, substitution.replace(/\$<([^>]+)>/g, function (_, name) { return "$" + groups[name]; })); } if ("function" == typeof substitution) { var _this = this; return _super[Symbol.replace].call(this, str, function () { var args = arguments; return "object" != _typeof(args[args.length - 1]) && (args = [].slice.call(args)).push(buildGroups(args, _this)), substitution.apply(this, args); }); } return _super[Symbol.replace].call(this, str, substitution); }, _wrapRegExp.apply(this, arguments); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var FLAT = 'b';
var SHARP = '#';
var MODIFIER_TRANSPOSITION = (_MODIFIER_TRANSPOSITI = {}, _defineProperty(_MODIFIER_TRANSPOSITI, SHARP, 1), _defineProperty(_MODIFIER_TRANSPOSITI, FLAT, -1), _MODIFIER_TRANSPOSITI);

var symbolKeyRegex = /*#__PURE__*/_wrapRegExp(/^([A-G])(#|b)?(m)?$/i, {
  note: 1,
  modifier: 2,
  minor: 3
});

var numericKeyRegex = /*#__PURE__*/_wrapRegExp(/^(#|b)?([1-7])(m)?$/, {
  modifier: 1,
  note: 2,
  minor: 3
});

var numeralKeyRegex = /*#__PURE__*/_wrapRegExp(/^(#|b)?(I{1,3}|IV|VI{0,2}|i{1,3}|iv|vi{0,2})$/, {
  modifier: 1,
  note: 2
});

var regexes = [symbolKeyRegex, numericKeyRegex, numeralKeyRegex];

var _set = /*#__PURE__*/new WeakSet();

var Key = /*#__PURE__*/function () {
  function Key(_ref) {
    var note = _ref.note,
        _ref$modifier = _ref.modifier,
        modifier = _ref$modifier === void 0 ? null : _ref$modifier,
        _ref$minor = _ref.minor,
        minor = _ref$minor === void 0 ? false : _ref$minor;

    _classCallCheck(this, Key);

    _classPrivateMethodInitSpec(this, _set);

    this.note = note instanceof _note["default"] ? note : _note["default"].parse(note);
    this.modifier = modifier || null;
    this.minor = !!minor || false;
  }

  _createClass(Key, [{
    key: "distanceTo",
    value: function distanceTo(otherKey) {
      var otherKeyObj = Key.wrap(otherKey);
      var key = this.useModifier(otherKeyObj.modifier);
      var delta = 0;

      while (!key.equals(otherKeyObj) && delta < 20) {
        key = key.transposeUp().useModifier(otherKeyObj.modifier);
        delta += 1;
      }

      return delta;
    }
  }, {
    key: "isMinor",
    value: function isMinor() {
      return this.note.isMinor();
    }
  }, {
    key: "clone",
    value: function clone() {
      return _classPrivateMethodGet(this, _set, _set2).call(this, {});
    }
  }, {
    key: "toChordSymbol",
    value: function toChordSymbol(key) {
      if (this.is(_constants.SYMBOL)) {
        return this.clone();
      }

      var transposeDistance = this.note.getTransposeDistance(key.minor) + (MODIFIER_TRANSPOSITION[this.modifier] || 0);
      return key.transpose(transposeDistance).normalize().useModifier(key.modifier);
    }
  }, {
    key: "toChordSymbolString",
    value: function toChordSymbolString(key) {
      return this.toChordSymbol(key).toString();
    }
  }, {
    key: "is",
    value: function is(type) {
      return this.note.is(type);
    }
  }, {
    key: "isNumeric",
    value: function isNumeric() {
      return this.is(_constants.NUMERIC);
    }
  }, {
    key: "isChordSymbol",
    value: function isChordSymbol() {
      return this.is(_constants.SYMBOL);
    }
  }, {
    key: "isNumeral",
    value: function isNumeral() {
      return this.is(_constants.NUMERAL);
    }
  }, {
    key: "equals",
    value: function equals(_ref2) {
      var note = _ref2.note,
          modifier = _ref2.modifier;
      return this.note.equals(note) && this.modifier === modifier;
    }
  }, {
    key: "toNumeric",
    value: function toNumeric(key) {
      if (this.isNumeric()) {
        return this.clone();
      }

      if (this.isNumeral()) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.toNumeric()
        });
      }

      var numericKey = new Key({
        note: 1
      });
      var symbolKey = key.clone();
      var reference = this.clone().normalize().useModifier(key.modifier).normalizeEnharmonics(key);

      while (!symbolKey.equals(reference)) {
        numericKey = numericKey.transposeUp().useModifier(key.modifier);
        symbolKey = symbolKey.transposeUp().normalize().useModifier(key.modifier).normalizeEnharmonics(key);
      }

      return numericKey;
    }
  }, {
    key: "toNumericString",
    value: function toNumericString(key) {
      return this.toNumeric(key).toString();
    }
  }, {
    key: "toNumeral",
    value: function toNumeral(key) {
      if (this.isNumeral()) {
        return this.clone();
      }

      if (this.isNumeric()) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.toNumeral()
        });
      }

      var numeralKey = new Key({
        note: 'I'
      });
      var symbolKey = key.clone();
      var reference = this.clone().normalize().useModifier(key.modifier).normalizeEnharmonics(key);

      while (!symbolKey.equals(reference)) {
        numeralKey = numeralKey.transposeUp().useModifier(key.modifier);
        symbolKey = symbolKey.transposeUp().normalize().useModifier(key.modifier).normalizeEnharmonics(key);
      }

      return numeralKey;
    }
  }, {
    key: "toNumeralString",
    value: function toNumeralString(key) {
      return this.toNumeral(key).toString();
    }
  }, {
    key: "toString",
    value: function toString() {
      if (this.isChordSymbol()) {
        return "".concat(this.note).concat(this.modifier || '');
      }

      return "".concat(this.modifier || '').concat(this.note);
    }
  }, {
    key: "transpose",
    value: function transpose(delta) {
      if (delta === 0) {
        return this;
      }

      var originalModifier = this.modifier;
      var transposedKey = this.clone();
      var func = delta < 0 ? 'transposeDown' : 'transposeUp';

      for (var i = 0, count = Math.abs(delta); i < count; i += 1) {
        transposedKey = transposedKey[func]();
      }

      return transposedKey.useModifier(originalModifier);
    }
  }, {
    key: "transposeUp",
    value: function transposeUp() {
      if (this.modifier === FLAT) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          modifier: null
        });
      }

      if (this.note.isOneOf(3, 7, 'E', 'B')) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.up()
        });
      }

      if (this.modifier === SHARP) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.up(),
          modifier: null
        });
      }

      return _classPrivateMethodGet(this, _set, _set2).call(this, {
        modifier: SHARP
      });
    }
  }, {
    key: "transposeDown",
    value: function transposeDown() {
      if (this.modifier === SHARP) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          modifier: null
        });
      }

      if (this.note.isOneOf(1, 4, 'C', 'F')) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.down()
        });
      }

      if (this.modifier === FLAT) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.down(),
          modifier: null
        });
      }

      return _classPrivateMethodGet(this, _set, _set2).call(this, {
        modifier: FLAT
      });
    }
  }, {
    key: "useModifier",
    value: function useModifier(newModifier) {
      if (this.modifier === FLAT && newModifier === SHARP) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.down(),
          modifier: SHARP
        });
      }

      if (this.modifier === SHARP && newModifier === FLAT) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.up(),
          modifier: FLAT
        });
      }

      return this.clone();
    }
  }, {
    key: "normalize",
    value: function normalize() {
      if (this.modifier === SHARP && this.note.isOneOf(3, 7, 'E', 'B')) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.up(),
          modifier: null
        });
      }

      if (this.modifier === FLAT && this.note.isOneOf(1, 4, 'C', 'F')) {
        return _classPrivateMethodGet(this, _set, _set2).call(this, {
          note: this.note.down(),
          modifier: null
        });
      }

      return this.clone();
    }
  }, {
    key: "normalizeEnharmonics",
    value: function normalizeEnharmonics(key) {
      if (key) {
        var rootKeyString = key.minor ? "".concat(key, "m") : key.toString();
        var enharmonics = _enharmonicNormalize["default"][rootKeyString];
        var thisKeyString = this.toString();

        if (enharmonics && enharmonics[thisKeyString]) {
          return Key.parse(enharmonics[thisKeyString]);
        }
      }

      return this.clone();
    }
  }], [{
    key: "parse",
    value: function parse(keyString) {
      for (var i = 0, count = regexes.length; i < count; i += 1) {
        var match = keyString.match(regexes[i]);

        if (match) {
          return new Key(match.groups);
        }
      }

      return null;
    }
  }, {
    key: "wrap",
    value: function wrap(keyStringOrObject) {
      if (keyStringOrObject instanceof Key) {
        return keyStringOrObject;
      }

      return this.parse(keyStringOrObject);
    }
  }, {
    key: "distance",
    value: function distance(oneKey, otherKey) {
      return this.wrap(oneKey).distanceTo(otherKey);
    }
  }]);

  return Key;
}();

function _set2(attributes) {
  return new this.constructor(_objectSpread({
    note: this.note.clone(),
    modifier: this.modifier,
    minor: !!this.minor
  }, attributes));
}

var _default = Key;
exports["default"] = _default;
